src/ObjectUtils/LinuxMap.cpp

```diff
@@ -89,32 +89,178 @@ ErrorMessageOr<std::vector<ModuleInfo>> ReadModules(int32_t pid) {
   return ParseMaps(proc_maps_data);
 }
 
+namespace {
+
+// If the .text section of a Portable Executable has an offset in the file (PointerToRawData,
+// multiple of FileAlignment) that is not congruent to the offset of that section when loaded into
+// memory (VirtualAddress, multiple of SectionAlignment) modulo the page size, Wine cannot create a
+// file-backed mapping for the .text section. This is because of the requirements on the arguments
+// of mmap. In such a case, Wine resorts to creating an anonymous mapping and copying the .text
+// section into it. This means that, for PE binaries with this property, we cannot simply associate
+// an executable mapping to the corresponding file using the path in the mapping.
+//
+// However, we can make an educated guess. The path of the PE will at least appear in the read-only
+// mapping that corresponds to the beginning of the file, which contains the headers (because the
+// offset in the file is zero and the address chosen for this mapping should always be multiple of
+// the page size). If the executable file mapping for the .text section is not present, we consider
+// the anonymous executable mappings after the first file mapping for this PE: if the offset and
+// size of such a mapping are compatible with the address range where the .text section would be
+// loaded based on the header for the section (in particular, VirtualAddress and VirtualSize), we
+// can be quite sure that this is the mapping we are looking for.
+// Note that we assume that a PE (or an ELF file) only has one .text section and one executable
+// mapping: this is what we observed and is what we support.
+//
+// This class contains logic to help ParseMaps with this detection mechanism. The intended usage is
+// as follows:
+// - Create a new instance of this class when a new file is encountered while parsing
+//   `/proc/[pid]/maps`;
+// - Call `MarkExecutableMapEncountered` when encountering an executable file mapping for the file
+//   this instance was created for.
+// - Use `TryIfAnonExecMapIsCoffTextSection` to query if an anonymous executable mapping is actually
+//   the PE .text section of the file this instance was created for. This will only return true once
+//   for each instance of this class, as it calls `MarkExecutableMapEncountered` on success.
+class FileMappedIntoMemory {
+ public:
+  FileMappedIntoMemory(std::string file_path, uint64_t first_map_start, uint64_t first_map_offset)
+      : file_path_{std::move(file_path)}, base_address_{first_map_start - first_map_offset} {}
```

Comment:
I didn't think about it.
I think this will always hold for regular inputs, because there's a minimum address below which mappings cannot be created (`/proc/sys/vm/mmap_min_addr`). But I don't think it's an absolute requirement, so conceptually I don't think we should add a check.
Additionally, `base_address_` is always summed to `cached_coff_file_->GetExecutableSegmentOffset()` before it's used in comparisons, and that should "undo" the underflow. Again, I don't think this is true in absolute terms.
What I've done is: in case of underflow, already set `coff_text_section_map_might_be_encountered_ = false;` and `base_address_` will not be used.
Let me know if you prefer a different solution.

Reply:
LGTM, thanks